nemo_skills/inference/model/tool_call.py

```diff
@@ -113,33 +114,38 @@ async def generate_async(
         result_steps = defaultdict(list)
         conversation = copy.deepcopy(prompt)
 
+        # assigning a unique request id to pass to tool calls if they need to be stateful
+        request_id = str(uuid.uuid4())
+
         while True:
             if isinstance(tokens_to_generate, int) and tokens_to_generate <= 0:
                 break
-
             generation = await self.model.generate_async(
                 prompt=conversation,
                 tools=tools,
                 tokens_to_generate=tokens_to_generate,
                 **generation_kwargs,
             )
-
             if isinstance(tokens_to_generate, int):
                 tokens_to_generate -= generation["num_generated_tokens"]
 
             for k in ["generation", "num_generated_tokens", "reasoning_content", "finish_reason"]:
                 if k in generation:
                     result_steps[k].append(generation[k])
 
-            conversation.append({"role": "assistant", "content": result_steps["generation"][-1]})
+            conversation.append({"role": "assistant", "content": generation["generation"]})
+            if "reasoning_content" in generation:
+                conversation[-1]["reasoning_content"] = generation["reasoning_content"]
 
             tool_calls = generation.get("tool_calls", [])
             if tool_calls:
                 tool_calls_message = self.call_interpreter.parse(tool_calls)
-                conversation.append(tool_calls_message)
+                conversation[-1].update(tool_calls_message)
```

Comment:
In Qwen3 at least, these are supposed to be a separate message, with "role": "tool", and "content" with the response of the tool. Is it different for some other model, do we need to encapsulate that?

Reply:
ah I see, yeah that's right